src/coap_gnutls.c

```diff
@@ -690,7 +717,7 @@ static int cert_verify_gnutls(gnutls_session_t g_session)
     G_CHECK(gnutls_x509_crt_export(cert, GNUTLS_X509_FMT_DER, der, &size),
             "gnutls_x509_crt_export");
     gnutls_x509_crt_deinit(cert);
-    if (!g_context->setup_data.validate_cn_call_back(cn,
+    if (!g_context->setup_data.validate_cn_call_back(cn ? cn : "RPK",
```

Comment:
I am not sure about it but I see a potential issue with treating an empty `cn` as RPK: The function `get_san_or_cn()` called in line 657 to set `cn` also returns NULL in case the certificate is of type `GNUTLS_CRT_X509` but the peer certificate list is empty or the certificate is malformed. In line 705 the certificate list is checked again. It might be worth differentiating explicitly between full X.509 certificates and raw public keys in `get_san_or_cn()`.  Maybe `gnutls_certificate_type_get2()` can help with this as we already require at least version 3.6.6 for RPKs. This function should return `GNUTLS_CRT_RAW` for the RPK. `get_san_or_cn()` then probably would look something like:

```c
typedef enum coap_gnutls_crt_t {
  COAP_GNUTLS_CRT_ERROR = -1,
  COAP_GNUTLS_CRT_X509 = 1,   /* GNUTLS_CRT_X509  */
  COAP_GNUTLS_CRT_RAW  = 3,   /* GNUTLS_CRT_RAW */
} coap_gnutls_crt_t;

typedef struct {
  coap_gnutls_crt_t certificate_type;
  char *san_or_cn;
  const gnutls_datum_t *cert_list;
  size_t cert_list_size;
} coap_gnutls_certificate_info_t;

static coap_gnutls_crt
get_san_or_cn(gnutls_session_t g_session, coap_gnutls_certificate_info_t *cert_info) {
...
#if (GNUTLS_VERSION_NUMBER >= 0x030304)
    gnutls_certificate_type_get2(...)
#else /* GNUTLS_VERSION_NUMBER < 0x030304 */
    gnutls_certificate_type_get(...)
#endif /* GNUTLS_VERSION_NUMBER < 0x030304 */
    cert_info->certificate_type = ...
   ...
   ... set cert_list from gnutls_certificate_get_peers() to avoid calling this function twice ...
}
```

Reply:
Code updated following the above guidelines with a couple of changes.